iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/stack.inl

```diff
@@ -22,6 +22,119 @@ namespace iox
 {
 namespace cxx
 {
+// AXIVION Next Construct AutosarC++19_03-A12.6.1 : the remaining m_data fields are explicitly initialized when a new
+// element is pushed
+// NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init, hicpp-member-init)
+template <typename T, uint64_t Capacity>
+inline stack<T, Capacity>::stack(const stack& rhs) noexcept
+{
+    *this = rhs;
+}
+
+// AXIVION Next Construct AutosarC++19_03-A12.6.1 : the remaining m_data fields are explicitly initialized when a new
+// element is pushed
+// NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init, hicpp-member-init)
+template <typename T, uint64_t Capacity>
+inline stack<T, Capacity>::stack(stack&& rhs) noexcept
+{
+    *this = std::move(rhs);
+}
+
+template <typename T, uint64_t Capacity>
+inline stack<T, Capacity>& stack<T, Capacity>::operator=(const stack& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        uint64_t i{0};
+        const uint64_t rhsSize{rhs.size()};
+        const uint64_t minSize{algorithm::minVal(m_size, rhsSize)};
+
+        // copy assignment
+        for (; i < minSize; i++)
+        {
+            getUnchecked(i) = rhs.getUnchecked(i);
+        }
+        // copy c'tor
+        for (; i < rhsSize; i++)
+        {
+            // AXIVION Next Line AutosarC++19_03-A18.5.2 : false positive, it is a placement new
+            new (&m_data[i]) T(rhs.getUnchecked(i));
+        }
+        // delete remaining elements
+        for (; i < m_size; i++)
+        {
+            getUnchecked(i).~T();
+        }
+
+        m_size = rhsSize;
+    }
+    return *this;
+}
+
+template <typename T, uint64_t Capacity>
+inline stack<T, Capacity>& stack<T, Capacity>::operator=(stack&& rhs) noexcept
+{
+    if (this != &rhs)
+    {
+        uint64_t i{0};
+        const uint64_t rhsSize{rhs.size()};
+        const uint64_t minSize{algorithm::minVal(m_size, rhsSize)};
+
+        // move assignment
+        for (; i < minSize; i++)
+        {
+            getUnchecked(i) = std::move(rhs.getUnchecked(i));
+        }
+        // move c'tor
+        for (; i < rhsSize; i++)
+        {
+            // AXIVION Next Line AutosarC++19_03-A18.5.2 : false positive, it is a placement new
+            new (&m_data[i]) T(std::move(rhs.getUnchecked(i)));
+        }
+        // delete remaining elements
+        for (; i < m_size; i++)
+        {
+            getUnchecked(i).~T();
+        }
+
+        m_size = rhsSize;
+
+        // clear rhs
+        for (uint64_t j{0}; j < rhsSize; j++)
```

Comment:
Could you please introduce a public method called clear which calls the destructor of all elements and sets the size to zero and call this method in the copy and move assignment as well as the destructor.

Furthermore, we have to remove the elements in revert direction from `m_size - 1` ... `0`.
When you for instance declare 3 variables or have 3 members a, b and c like:
```
int a;
int b;
int c;
```

The destruction order is always in the opposite creation order, so first `c` is destructed then `b` and then `a`. It is strange but possible for instance that element `c` inside the stack has a reference to `b` and requires that reference during destruction. This would be completely legal since the stack resides on the heap and never relocates.

With the current stack destruction behavior we would have a problem.

Furthermore, see the C++ standard: https://isocpp.org/files/papers/N4860.pdf, section 7.6.2.8 Delete, subsection 6 (page 129)
```
In the case of an array, the elements will
be destroyed in order of decreasing address (that is, in reverse order of the completion of their constructor;
see 11.10.2).
```

Fun fact, when I compiled this piece of code:
```cxx
#include <iostream>
#include <stack>
#include <vector>

struct A {
    A(int a) : a{a} { std::cout << "create: " << a << std::endl; }
    ~A() { std::cout << "destroy: " << a << std::endl; }

    int a;
};

// namespace iox
int main() {
    std::vector<A> a;
    a.push_back(A(1));
    a.push_back(A(2));
    a.push_back(A(3));
}
```

It showed me that my STL implementation does not destroy the elements in reverse order. (Same goes when you replace `std::vector` with `std::stack`). This is maybe a bug.

Reply:
When we agree that this has to be adjusted we also would require a unit tests which verifies the reverse destruction order.
@elBoberido what do you think?

Reply:
When I am right we have another funky problem since we are doing it wrong in every container.

Reply:
@elfenpiff I think you are right and it should be done in reverse order, at least for container where the items are stored consecutively. Not sure what to do in a `list`. A double linked list could also be destroyed in reverse order but with a single linked list this is hard to achieve

@MatthiasKillat what do you think?

Reply:
@elfenpiff I think you're right. I will adapt the destruction order here, check our other containers and create an issue.

Reply:
I created an issue for `cxx::vector`: #1643. The other containers seem to be alright, although the destruction order for the lists may be debatable.